### PR TITLE
Cutomized alpn

### PR DESCRIPTION
**Pull request by TingDaoK**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/connection.c [hunk 1]
```diff
@@ -963,6 +1051,27 @@ int aws_http_client_connect_internal(
         http_bootstrap->http2_options.initial_settings_array = setting_array;
     }
 
+    if (options.alpn_string_map) {
+        alpn_string_map = aws_http_default_alpn_map_new(options.allocator);
```
Comment (graebm) [inline source/connection.c:1055]: I vote that we be explicit when the user passes in an ALPN map: Copy it exactly, and reject anything that's not in the list.

It's a little weird to automatically insert "h2" and "http/1.1" when the user didn't ask. It's also a little weird to fall back to HTTP/1.1 if the negotiated string doesn't match anything. We added the fallback to HTTP/1.1 when we needed to support a custom ALPN string, but had no controls for doing so. But now we DO have a control for doing so.

File: tests/test_connection.c [hunk 2]
```diff
@@ -548,6 +553,77 @@ static int s_test_connection_h2_prior_knowledge_not_work_with_tls(struct aws_all
...
+    ASSERT_SUCCESS(aws_http_client_connect(&tester.client_options));
+    /* We should be safe to free the map */
+    aws_hash_table_clean_up(alpn_map);
+
+    /* Wait for server & client connections to finish setup */
+    tester.wait_client_connection_num = 1;
+    tester.wait_server_connection_num = 1;
+    ASSERT_SUCCESS(s_tester_wait(&tester, s_tester_connection_setup_pred));
+
+#ifndef __APPLE__ /* Server side ALPN doesn't work for MacOS */
```
Comment (graebm) [inline tests/test_connection.c:612]: you can remove this `ifndef`, you already have an `ifdef __APPLE__` earlier in the function

File: tests/test_connection.c [hunk 3]
```diff
@@ -548,6 +553,77 @@ static int s_test_connection_h2_prior_knowledge_not_work_with_tls(struct aws_all
...
+    struct aws_byte_buf src = aws_tls_handler_protocol(handler);
+    aws_byte_buf_init_copy(tester->negotiated_protocol, tester->alloc, &src);
+}
+
+static int s_test_connection_customized_alpn(struct aws_allocator *allocator, void *ctx) {
+    (void)ctx;
+#ifdef __APPLE__ /* Something is wrong with APPLE */
+    return AWS_OP_SUCCESS;
+#endif
+    char customized_alpn_string[] = "myh2";
```
Comment (graebm) [inline tests/test_connection.c:574]: you should probably put the rest of this code in an `#else` block
won't the compiler complain about code that's impossible to reach after the `return`?

File: source/connection.c (hunk 1 above)
Comment (graebm) [reply source/connection.c:1055]: oh. I was confused. you ARE copying it exactly.

I was thrown off by the word "default" in `aws_http_default_alpn_map_new()`
How about just `aws_http_alpn_map_new()`

File: include/aws/http/connection.h [hunk 4]
```diff
@@ -433,6 +443,14 @@ enum aws_http_version aws_http_connection_get_version(const struct aws_http_conn
 AWS_HTTP_API
 struct aws_channel *aws_http_connection_get_channel(struct aws_http_connection *connection);
 
+/**
+ * Helper function that returns a default map can be used in aws_http_client_connection_options as alpn_string_map.
+ * The caller owns the returned map.
+ * The aws_string created as the key of the map will be owned by the map.
+ */
+AWS_HTTP_API
+struct aws_hash_table *aws_http_default_alpn_map_new(struct aws_allocator *allocator);
```
[the remark this answers is not in the document]
Comment (graebm) [reply include/aws/http/connection.h:452]: Change comments and name. There's nothing "default" about this. It returns an empty map.

It's ok for this to stay in a public header. I see how this is useful.

File: source/connection.c [hunk 5]
```diff
@@ -124,7 +135,24 @@ struct aws_http_connection *aws_http_connection_new_channel_handler(
...
         if (protocol.len) {
-            if (aws_string_eq_byte_buf(s_alpn_protocol_http_1_1, &protocol)) {
+            bool customized = false;
+            if (alpn_string_map) {
+                struct aws_string *negotiated_result = aws_string_new_from_buf(alloc, &protocol);
+                struct aws_hash_element *found = NULL;
+                aws_hash_table_find(alpn_string_map, (void *)negotiated_result, &found);
+                if (found) {
+                    version = (enum aws_http_version)found->value;
+                    customized = true;
```
Comment (graebm) [inline source/connection.c:145]: if `found` is `false` we're going to end up falling back to HTTP/1.1
probably want it to be an error if `found` is false

File: tests/test_connection.c (hunk 2 above)
Comment (TingDaoK) [reply tests/test_connection.c:612]: The first `ifdef __APPLE__` should be removed as long as the github virtual env solves the [issue](https://github.com/actions/virtual-environments/issues/3861). It's expected to be fixed this Thursday.

File: tests/test_connection.c (hunk 3 above)
Comment (TingDaoK) [reply tests/test_connection.c:574]: Don't know... But, I'll remove this block after we have the Mac virtual env back to work. We can ignore it for now. (I can keep this PR open until then)

File: source/connection.c [hunk 6]
```diff
@@ -885,6 +945,37 @@ static int s_validate_http_client_connection_options(const struct aws_http_clien
     return AWS_OP_SUCCESS;
 }
 
+struct s_copy_alpn_string_map_context {
+    struct aws_hash_table *map;
+    struct aws_allocator *allocator;
+};
```
[the remark this answers is not in the document]
Comment (TingDaoK) [reply source/connection.c:951]: because aws_hash_table doesn't have a allocator in it...

File: include/aws/http/connection.h [hunk 7]
```diff
@@ -452,6 +462,14 @@ enum aws_http_version aws_http_connection_get_version(const struct aws_http_conn
 AWS_HTTP_API
 struct aws_channel *aws_http_connection_get_channel(struct aws_http_connection *connection);
 
+/**
+ * Helper function that returns an empty map can be used in aws_http_client_connection_options as alpn_string_map.
+ * The caller owns the returned map.
+ * The aws_string created as the key of the map will be owned by the map.
+ */
+AWS_HTTP_API
+int aws_http_alpn_map_init(struct aws_allocator *allocator, struct aws_hash_table *map);
```
Comment (graebm) [inline include/aws/http/connection.h:471]: Not sure what you were trying to say with "The aws_string created as the key of the map will be owned by the map." ??

how about just:
```
/**
 * Initialize an empty hash-table that maps `struct aws_string *` to `enum aws_http_version`.
 * This map can used in aws_http_client_connections_options.alpn_string_map.
 */
```

File: source/connection.c [hunk 8]
```diff
@@ -31,6 +31,15 @@ static struct aws_http_connection_system_vtable s_default_system_vtable = {
 
 static const struct aws_http_connection_system_vtable *s_system_vtable_ptr = &s_default_system_vtable;
 
+void aws_http_client_bootstrap_destroy(struct aws_http_client_bootstrap *bootstrap) {
+    /* During allocating, the underlying stuctures should be allocated with the bootstrap by aws_mem_acquire_many. Thus,
+     * we only need to clean up the first pointer which is the boostrap */
```
[the remark this answers is not in the document]
Comment (graebm) [reply source/connection.c:36]: BOOSTRAP

**Review by graebm — APPROVED** [review]
comment tweaks and shipit

File: source/connection.c [hunk 9]
```diff
@@ -959,6 +1057,26 @@ int aws_http_client_connect_internal(
...
+        /* make a deep copy of the map */
+        if (aws_hash_table_foreach(options.alpn_string_map, s_copy_alpn_string_map, &context)) {
+            int error_code = aws_last_error();
+            AWS_LOGF_ERROR(
+                AWS_LS_HTTP_CONNECTION,
+                "Failed to copy ALPN map with error code %d (%s)",
+                error_code,
+                aws_error_name(error_code));
+            goto error;
+        }
```
[the remark this answers is not in the document]
Comment (TingDaoK) [reply source/connection.c]: changed to `aws_http_alpn_map_init_copy`

**Pull request merged**